### ISSUE TITLE
error validating btree node at btree 4 level 0/1

### ISSUE THREAD

**Issue by marcin-github** [koverstreet/bcachefs#77]
```
[   13.592296] bcachefs (sdc1): journal read done, 4346274 keys in 1026 entries, seq 245850
[   25.255591] bcachefs (sdc1): error validating btree node at btree 4 level 0/1
               pos 0:5036935 node offset 0 bset u64s 23498: incorrect max key
[   25.255596] bcachefs (sdc1): error reading alloc info: -5
[   25.311529] bcachefs (sdc1): Error in recovery: error reading allocation information (-5)
```
`mount  -o reconstruct_alloc,fix_errors ` can't fix it.

**Comment by nicman23** [comment]
try fsck, fix_errors

**Comment by marcin-github** [comment]
Tried without luck. I wiped partition because I couldn't fix it.

**Comment by hyperfekt** [comment]
```
Apr 05 19:03:12 nixos sudo[1603]:    nixos : TTY=pts/1 ; PWD=/home/nixos ; USER=root ; COMMAND=/run/current-system/sw/bin/mount -t bcachefs /dev/sdb1:/dev/sda2 /mnt
Apr 05 19:03:12 nixos sudo[1603]: pam_unix(sudo:session): session opened for user root by (uid=0)
Apr 05 19:03:12 nixos kernel: raid6: avx2x4   gen() 32957 MB/s
Apr 05 19:03:13 nixos kernel: raid6: avx2x4   xor() 20951 MB/s
Apr 05 19:03:13 nixos kernel: raid6: avx2x2   gen() 30046 MB/s
Apr 05 19:03:13 nixos kernel: raid6: avx2x2   xor() 19410 MB/s
Apr 05 19:03:13 nixos kernel: raid6: avx2x1   gen() 23378 MB/s
Apr 05 19:03:13 nixos kernel: raid6: avx2x1   xor() 16072 MB/s
Apr 05 19:03:13 nixos kernel: raid6: sse2x4   gen() 14148 MB/s
Apr 05 19:03:13 nixos kernel: raid6: sse2x4   xor()  8484 MB/s
Apr 05 19:03:13 nixos kernel: raid6: sse2x2   gen() 12238 MB/s
Apr 05 19:03:13 nixos kernel: raid6: sse2x2   xor()  8189 MB/s
Apr 05 19:03:13 nixos kernel: raid6: sse2x1   gen() 10929 MB/s
Apr 05 19:03:13 nixos kernel: raid6: sse2x1   xor()  6035 MB/s
Apr 05 19:03:13 nixos kernel: raid6: using algorithm avx2x4 gen() 32957 MB/s
Apr 05 19:03:13 nixos kernel: raid6: .... xor() 20951 MB/s, rmw enabled
Apr 05 19:03:13 nixos kernel: raid6: using avx2x2 recovery algorithm
Apr 05 19:03:13 nixos kernel: xor: automatically using best checksumming function   avx       
Apr 05 19:03:19 nixos kernel: bcachefs (24a3914b-3848-4b47-a36f-7ef57591f683): journal read done, 0 keys in 1 entries, seq 5488092
Apr 05 19:03:20 nixos kernel: bcachefs (24a3914b-3848-4b47-a36f-7ef57591f683): starting metadata mark and sweep
Apr 05 19:03:23 nixos kernel: bcachefs (24a3914b-3848-4b47-a36f-7ef57591f683): error validating btree node at btree 2 level 0/1
                              pos 1409330:5324218192610098858 node offset 0: bad magic
Apr 05 19:03:23 nixos kernel: bcachefs (24a3914b-3848-4b47-a36f-7ef57591f683): error in fsck: error -5 while deleting inode
Apr 05 19:03:23 nixos kernel: bcachefs (24a3914b-3848-4b47-a36f-7ef57591f683): Error in recovery: error in recovery (-5)
Apr 05 19:03:24 nixos sudo[1603]: pam_unix(sudo:session): session closed for user root
Apr 05 19:03:43 nixos sudo[1638]:    nixos : TTY=pts/1 ; PWD=/home/nixos ; USER=root ; COMMAND=/run/current-system/sw/bin/mount -t bcachefs -o fsck,fix_errors /dev/sdb1:/dev/sda2 /mnt
Apr 05 19:03:43 nixos sudo[1638]: pam_unix(sudo:session): session opened for user root by (uid=0)
Apr 05 19:03:50 nixos kernel: bcachefs (24a3914b-3848-4b47-a36f-7ef57591f683): journal read done, 0 keys in 1 entries, seq 5488106
Apr 05 19:03:51 nixos kernel: bcachefs (24a3914b-3848-4b47-a36f-7ef57591f683): starting metadata mark and sweep
Apr 05 19:03:51 nixos kernel: bcachefs (24a3914b-3848-4b47-a36f-7ef57591f683): starting mark and sweep
Apr 05 19:04:09 nixos kernel: bcachefs (24a3914b-3848-4b47-a36f-7ef57591f683): bucket 0:11046 data type cached ptr gen in the future: 4 > 142, fixing
Apr 05 19:04:09 nixos kernel: bcachefs (24a3914b-3848-4b47-a36f-7ef57591f683): bucket 0:83890 data type cached ptr gen in the future: 40 > 178, fixing
Apr 05 19:04:09 nixos kernel: bcachefs (24a3914b-3848-4b47-a36f-7ef57591f683): bucket 0:126602 data type cached ptr gen in the future: 8 > 154, fixing
Apr 05 19:04:09 nixos kernel: bcachefs (24a3914b-3848-4b47-a36f-7ef57591f683): bucket 0:126602 gen 8 different types of data in same bucket: btree, data (run fsck to correct)
Apr 05 19:04:09 nixos kernel: bcachefs (24a3914b-3848-4b47-a36f-7ef57591f683): bucket 0:182459 data type cached ptr gen in the future: 53 > 182, fixing
Apr 05 19:04:09 nixos kernel: bcachefs (24a3914b-3848-4b47-a36f-7ef57591f683): Fixed gens, restarting mark and sweep:
Apr 05 19:04:09 nixos kernel: bcachefs (24a3914b-3848-4b47-a36f-7ef57591f683): bucket 0:182459 data type cached ptr gen in the future: 180 > 53, fixing
Apr 05 19:04:11 nixos kernel: bcachefs (24a3914b-3848-4b47-a36f-7ef57591f683): bucket 0:182459 data type cached ptr gen in the future: 181 > 180, fixing
Apr 05 19:04:12 nixos kernel: bcachefs (24a3914b-3848-4b47-a36f-7ef57591f683): bucket 0:182459 data type cached ptr gen in the future: 182 > 181, fixing
Apr 05 19:04:15 nixos kernel: bcachefs (24a3914b-3848-4b47-a36f-7ef57591f683): bucket 0:182459 data type cached ptr gen in the future: 53 > 182, fixing
Apr 05 19:04:15 nixos kernel: bcachefs (24a3914b-3848-4b47-a36f-7ef57591f683): Fixed gens, restarting mark and sweep:
Apr 05 19:04:15 nixos kernel: bcachefs (24a3914b-3848-4b47-a36f-7ef57591f683): bucket 0:182459 data type cached ptr gen in the future: 180 > 53, fixing
Apr 05 19:04:16 nixos kernel: bcachefs (24a3914b-3848-4b47-a36f-7ef57591f683): bucket 0:182459 data type cached ptr gen in the future: 181 > 180, fixing
Apr 05 19:04:17 nixos kernel: bcachefs (24a3914b-3848-4b47-a36f-7ef57591f683): bucket 0:182459 data type cached ptr gen in the future: 182 > 181, fixing
Apr 05 19:04:20 nixos kernel: bcachefs (24a3914b-3848-4b47-a36f-7ef57591f683): bucket 0:182459 data type cached ptr gen in the future: 53 > 182, fixing
Apr 05 19:04:20 nixos kernel: bcachefs (24a3914b-3848-4b47-a36f-7ef57591f683): Fixed gens, restarting mark and sweep:
Apr 05 19:04:20 nixos kernel: bcachefs (24a3914b-3848-4b47-a36f-7ef57591f683): bucket 0:182459 data type cached ptr gen in the future: 180 > 53, fixing
Apr 05 19:04:22 nixos kernel: bcachefs (24a3914b-3848-4b47-a36f-7ef57591f683): bucket 0:182459 data type cached ptr gen in the future: 181 > 180, fixing
Apr 05 19:04:23 nixos kernel: bcachefs (24a3914b-3848-4b47-a36f-7ef57591f683): bucket 0:182459 data type cached ptr gen in the future: 182 > 181, fixing
Apr 05 19:04:26 nixos kernel: bcachefs (24a3914b-3848-4b47-a36f-7ef57591f683): bucket 0:182459 data type cached ptr gen in the future: 53 > 182, fixing
Apr 05 19:04:26 nixos kernel: bcachefs (24a3914b-3848-4b47-a36f-7ef57591f683): Unable to fix bucket gens, looping
Apr 05 19:04:26 nixos kernel: bcachefs (24a3914b-3848-4b47-a36f-7ef57591f683): Error in recovery: error in mark and sweep (-22)
```
[Dumps](https://drive.google.com/open?id=1KsRlsfYeQwkRwcSYnliSL5TW2g0bJwoT) [here](https://drive.google.com/open?id=1xp9mDj1FE18aE3uGHuKRbuSIVLtxYNCd).

**Comment by Lyamc** [comment]
This has happened to three of my drives.

```
$ sudo bcachefs fsck /dev/sdf1
journal read done, 4646 keys in 1 entries, seq 197688 error validating btree node at btree 4 level 0/1 pos u64s 6 type btree_ptr 0:8746920 snap 0 len 0 ver 0: ptr: 0:4432429568 gen 19 stale node offset 0 bset u64s 19304: incorrect max key error reading alloc info: -5 Error in recovery: error reading allocation information (-5) error opening /dev/sdf1: Input/output error
```

**Comment by Lyamc** [comment]
Here's what I get to one of them when I do sudo bcachefs fsck -o reconstruct_alloc /dev/sdi1 -y
`journal read done, 83204 keys in 12 entries, seq 220
starting mark and sweep
bucket 0:90534 data type user ptr gen 9 missing in alloc btree, fixing
bucket 0:90525 data type btree ptr gen 9 missing in alloc btree, fixing
bucket 0:1032 data type btree ptr gen 1 missing in alloc btree, fixing
bucket 0:1033 data type btree ptr gen 1 missing in alloc btree, fixing
bcachefs: libbcachefs/btree_update_leaf.c:492: do_bch2_trans_commit: Assertion `!(!btree_node_intent_locked(i->iter, i->iter->level))' failed.
`

**Comment by koverstreet** [comment]
Lyamc - can you take a dump for me?

**Comment by Lyamc** [comment]
No, it fails the dump:

`sudo bcachefs dump /dev/sdg1 -o sdg1.qcow2

journal read done, 4646 keys in 4 entries, seq 197703

error validating btree node at btree 4 level 0/1

pos u64s 6 type btree_ptr 0:8746920 snap 0 len 0 ver 0: ptr: 0:4432429568 gen 19 stale node offset 0 bset u64s 19304: incorrect max key 0:8688993

error reading alloc info: -5

Error in recovery: error reading allocation information (-5)

error opening /dev/sdg1: Input/output error
`

**Comment by koverstreet** [comment]
-o reconstruct_alloc is fixed now by 8436db7aac

Closing this for now as I _think_ I fixed the original bug awhile back - reopen if anyone sees it again
